os/net/app-layer/snmp/snmp.h

```diff
@@ -0,0 +1,220 @@
+ * @{
+ */
+
+/**
+ * @brief SNMP Version 1 code
+ */
+#define SNMP_VERSION_1  0
+/**
+ * @brief SNMP Version 2c code
+ */
+#define SNMP_VERSION_2C 1
+
+/**
+ * @brief SNMP No Such Name error code
+ */
+#define SNMP_STATUS_NO_SUCH_NAME 2
+
+/** @} */
+
+/**
+ * \defgroup SNMPStructs SNMP Structs
+ * @{
+ */
+
+/**
+ * @brief The SNMP header struct
+ */
+typedef struct snmp_header_s {
+  /**
+   * @brief SNMP Version
+   */
+  uint32_t version;
+  /**
+   * @brief Struct to wrap the community
+   */
+  struct snmp_msg_community {
+    /**
+     * @brief A pointer to the community
+     *
+     * @remarks This pointer refers to the beginning of the string in the packet
+     */
+    const char *community;
+    /**
+     * @brief The string length
+     *
+     * @remarks Do not use strlen on the community pointer since it is not null terminated
+     */
+    uint32_t length;
+  } community;
+  /**
+   * @brief The PDU type
+   */
+  uint8_t pdu_type;
+  /**
+   * @brief The request ID
+   */
+  uint32_t request_id;
+  /**
+   * @brief Union to hold the error status or the non repeaters
+   *
+   * @remarks A union was used since these values cannot co-exist
+   */
+  union error_status_non_repeaters_u {
+    /**
+     * @brief The error status
+     */
+    uint32_t error_status;
+    /**
+     * @brief The non repeaters
+     */
+    uint32_t non_repeaters;
+  } error_status_non_repeaters;
+  /**
+   * @brief Union to hold the error index or the max repetitions
+   *
+   * @remarks A union was used since these values cannot co-exist
+   */
+  union error_index_max_repetitions_u {
+    /**
+     * @brief The error index
+     */
+    uint32_t error_index;
+    /**
+     * @brief The max repetitions
+     */
+    uint32_t max_repetitions;
+  } error_index_max_repetitions;
+} snmp_header_t;
+
+/**
+ * @brief The varbind struct
+ */
+typedef struct snmp_varbind_s {
+  /**
+   * @brief The OID
+   *
+   * @remarks The length is configurable
+   */
+  uint32_t oid[SNMP_MSG_OID_MAX_LEN];
+  /**
+   * @brief The type in this varbind
+   */
+  uint8_t value_type;
+  /**
+   * @brief A union to represent the value in this varbind
+   *
+   * @remarks A union is used since the varbind can only have one value of one type
+   */
+  union snmp_varbind_val_u {
+    /**
+     * @brief The integer value
+     */
+    uint32_t integer;
+    /**
+     * @brief A struct that contains the string
+     */
+    struct snmp_varbind_string_s {
+      /**
+       * @brief A pointer to the string value from this varbind
+       *
+       * @remarks This pointer points to a string that cannot be changed
+       */
+      const char *string;
+      /**
+       * @brief The string length
+       *
+       * @remarks Do not use strlen on the string since it might not be null terminated
+       */
+      uint32_t length;
+    } string;
+    /**
+     * @brief A pointer to the beggining of a oid array
+     */
+    uint32_t *oid;
+  } value;
+} snmp_varbind_t;
```

Comment:
Both `snmp_header_t` and `snmp_varbind_t` are very likely to get padded by the compiler. Is this OK? I did not check how they are being used, but if they are used to create/parse packets such padding could create trouble.

Reply:
Padding isn't a problem. SNMP uses the BER encoding, so the structs are only holding the information before encoding or after decoding. I tried to avoid padding, but since I used unions, I have to admit that I forgot to reorder the struct.

Reply:
OK so this is just a question about total size then